### PR TITLE
Missing the "stepResults" in "lastInstallresult" of device twin json.

### PR DESCRIPTION
**Pull request by lillianxu-eaton**
It causes the deployment status shows "Failed" on second OTA trial.

### CODE REVIEW

**Comment by lillianxu-eaton** [comment]
@microsoft-github-policy-service agree company="Eaton Corporation"

From: microsoft-github-policy-service[bot] ***@***.***>
Sent: Thursday, November 30, 2023 9:10 AM
To: Azure/azure-sdk-for-c ***@***.***>
Cc: Xu, Lillian ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [Azure/azure-sdk-for-c] Missing the "stepResults" in "lastInstallresult" of device twin json. (PR #2701)


@lillianxu-eaton<https://github.com/lillianxu-eaton> the command you issued was incorrect. Please try again.

Examples are:

@microsoft-github-policy-service agree

and

@microsoft-github-policy-service agree company="your company"

-

File: sdk/src/azure/iot/az_iot_adu_client.c [hunk 1]
```diff
@@ -279,20 +284,11 @@ AZ_NODISCARD az_result az_iot_adu_client_get_agent_state_payload(
       _az_RETURN_IF_FAILED(az_json_writer_append_int32(
           ref_json_writer, last_install_result->step_results[i].extended_result_code));
 
-      if (!az_span_is_content_equal(
-              last_install_result->step_results[i].result_details, AZ_SPAN_EMPTY))
-      {
-        _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
-            ref_json_writer,
-            AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_RESULT_DETAILS)));
```
Comment (vaavva) [inline sdk/src/azure/iot/az_iot_adu_client.c:287]: Why would this section for the result details be removed?

**Comment by lillianxu-eaton** [comment]
Hi Valerie,

The Microsoft team advised us that the device twin should be below.

Here is how it should look:

                    "lastInstallResult": {
                        "extendedResultCode": 0,
                        "resultCode": 0,
                        "resultDetails": "Success",
                        "stepResults": {
                            "step_0": {
                                "extendedResultCode": 0,
                                "resultCode": 0
                            }
                        }
                    },

Therefore, the part below is removed. Otherwise, the buffer will be overflowed and the device will be reset.

***@***.***

Thanks

Lillian

From: Valerie Avva Lim ***@***.***>
Sent: Monday, December 4, 2023 7:18 PM
To: Azure/azure-sdk-for-c ***@***.***>
Cc: Xu, Lillian ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [Azure/azure-sdk-for-c] Missing the "stepResults" in "lastInstallresult" of device twin json. (PR #2701)


@vaavva commented on this pull request.

________________________________

In sdk/src/azure/iot/az_iot_adu_client.c<https://github.com/Azure/azure-sdk-for-c/pull/2701#discussion_r1414687396>:

> @@ -279,20 +284,11 @@ AZ_NODISCARD az_result az_iot_adu_client_get_agent_state_payload(

       _az_RETURN_IF_FAILED(az_json_writer_append_int32(

           ref_json_writer, last_install_result->step_results[i].extended_result_code));



-      if (!az_span_is_content_equal(

-              last_install_result->step_results[i].result_details, AZ_SPAN_EMPTY))

-      {

-        _az_RETURN_IF_FAILED(az_json_writer_append_property_name(

-            ref_json_writer,

-            AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_RESULT_DETAILS)));

Why would this section for the result details be removed?

-

**Comment by ericwol-msft** [comment]
@lillianxu-eaton 

> Therefore, the part below is removed. Otherwise, the buffer will be overflowed and the device will be reset.

Can you provide more details on this? **Buffer overflows should be checked by the json_writer APIs.**

Is there a callstack for when this happens?

**Comment by lillianxu-eaton** [comment]
Hi Eric,

There should be only two items in the step_0.
***@***.***

But the part below will add the "resultDetails" to device twin. Please see the screenshot below. This causes the device reset.

***@***.***

***@***.***

From: Eric Wolz ***@***.***>
Sent: Tuesday, December 5, 2023 2:34 PM
To: Azure/azure-sdk-for-c ***@***.***>
Cc: Xu, Lillian ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [Azure/azure-sdk-for-c] Missing the "stepResults" in "lastInstallresult" of device twin json. (PR #2701)

You don't often get email from ***@***.******@***.***>. Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>

@lillianxu-eaton<https://github.com/lillianxu-eaton>

Therefore, the part below is removed. Otherwise, the buffer will be overflowed and the device will be reset.

Can you provide more details on this? Buffer overflows should be checked by the json_writer APIs.

Is there a callstack for when this happens?

-

**Comment by lillianxu-eaton** [comment]
Hi Eric,

I reviewed the code and the resultDetails is in the step_result structure. See below screenshot for details.
***@***.***

The "result_details" in both "lastInstallResult" and "stepResults" are set to string "Success" after the firmware is updated successfully.
Could you please advise any potential issues?

Thanks

Lillian

From: Xu, Lillian
Sent: Tuesday, December 5, 2023 3:14 PM
To: Azure/azure-sdk-for-c ***@***.***>; Azure/azure-sdk-for-c ***@***.***>
Cc: Mention ***@***.***>
Subject: RE: [EXTERNAL] Re: [Azure/azure-sdk-for-c] Missing the "stepResults" in "lastInstallresult" of device twin json. (PR #2701)

Hi Eric,

There should be only two items in the step_0.
***@***.***

But the part below will add the "resultDetails" to device twin. Please see the screenshot below. This causes the device reset.

***@***.***

***@***.***

From: Eric Wolz ***@***.******@***.***>>
Sent: Tuesday, December 5, 2023 2:34 PM
To: Azure/azure-sdk-for-c ***@***.******@***.***>>
Cc: Xu, Lillian ***@***.******@***.***>>; Mention ***@***.******@***.***>>
Subject: [EXTERNAL] Re: [Azure/azure-sdk-for-c] Missing the "stepResults" in "lastInstallresult" of device twin json. (PR #2701)

You don't often get email from ***@***.******@***.***>. Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>

@lillianxu-eaton<https://github.com/lillianxu-eaton>

Therefore, the part below is removed. Otherwise, the buffer will be overflowed and the device will be reset.

Can you provide more details on this? Buffer overflows should be checked by the json_writer APIs.

Is there a callstack for when this happens?

-

**Comment by vaavva** [comment]
@lillianxu-eaton I have opened a separate pull request (https://github.com/Azure/azure-sdk-for-c/pull/2708) to include the necessary testing changes and align with the documented format. Please let me know if you have issues with the fix in that PR after increasing your buffer size.

**Pull request closed without merging**